### Technical references

Bolger AM, Lohse M, Usadel B. Trimmomatic: a flexible trimmer for illumina sequence data. Bioinformatics .
2014:30(15):2114–2120. https://doi.org/10.1093/bioinformatics/btu170.

Bonfield JK, McCarthy SA, Durbin R. Crumble: reference free lossy compression of sequence quality values.
Bioinformatics . 2019:35(2):337–339. https://doi.org/10.1093/bioinformatics/bty608.

Bouckaert RR. DensiTree: making sense of sets of phylogenetic trees. Bioinformatics . 2010:26(10):1372–1373.
https://doi.org/10.1093/bioinformatics/btq110.

Bredemeyer KR, Hillier L, Harris AJ, Hughes GM, Foley NM, Lawless C, Carroll RA, Storer JM, Batzer MA,
Rice ES, et al.. Single-haplotype comparative genomics provides insights into lineage-specific structural variation during cat evolution. Nature Genetics . 2023:55(11):1953–1963. https://doi.org/10.1038/s41588-023-01548-y.

Danecek P, Bonfield JK, Liddle J, Marshall J, Ohan V, Pollard MO, Whitwham A, Keane T, McCarthy SA,
Davies RM, et al. Twelve years of SAMtools and BCFtools. Gigascience . 2021:10(2):giab008.
https://doi.org/10.1093/gigascience/giab008.

Edelman NB, Frandsen PB, Miyagi M, Clavijo B, Davey J, Dikow RB, Garcia-Accinelli G, Van Belleghem SM,
Patterson N, Neafsey DE, et al. Genomic architecture and introgression shape a butterfly radiation. Science .
2019:366(6465):594–599. https://doi.org/10.1126/science.aaw2090.

Edgar RC. MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Res .
2004:32(5):1792–1797. https://doi.org/10.1093/nar/gkh340.

Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single
report. Bioinformatics . 2016:32(19):3047–3048. https://doi.org/10.1093/bioinformatics/btw354.

Green RE, Krause J, Briggs AW, Maricic T, Stenzel U, Kircher M, Patterson N, Li H, Zhai WW, Fritz MHY, et
al. A draft sequence of the neandertal genome. Science . 2010:328(5979):710–722. https://doi.org/10.1126/science.1188021.

Hahn C, Bachmann L, Chevreux B. Reconstructing mitochondrial genomes directly from genomic next-generation
sequencing reads—a baiting and iterative mapping approach. Nucleic Acids Res . 2013:41(13):e129.
https://doi.org/10.1093/nar/gkt371.

Huson DH, Bryant D. Application of phylogenetic networks in evolutionary studies. Mol Biol Evol. 2006:23(2):254–267.
https://doi.org/10.1093/molbev/msj030.

Jukes TH, Cantor CR. Evolution of Protein Molecules. In: Munro HN, editors. Mammalian Protein Metabolism. New York:
Academic Press; 1969. p. 21–132. 10.1016/B978-1-4832-3211-9.50009-7.

Junier T, Zdobnov EM. The newick utilities: high-throughput phylogenetic tree processing in the unix shell. Bioinformatics. 2010:26(13):1669–1670. https://doi.org/10.1093/bioinformatics/btq243.

Korneliussen TS, Albrechtsen A, Nielsen R. ANGSD: analysis of next generation sequencing data. BMC Bioinformatics .
2014:15(1):356. https://doi.org/10.1186/s12859-014-0356-4.

Kumar S, Stecher G, Tamura K. MEGA7: molecular evolutionary genetics analysis version 7.0 for bigger datasets. Mol Biol
Evol. 2016:33(7):1870–1874. https://doi.org/10.1093/molbev/msw054.

Li H. Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics . 2018:34(18):3094–3100.
https://doi.org/10.1093/bioinformatics/bty191.

Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R, Proc GPD. The
sequence alignment/map format and SAMtools. Bioinformatics . 2009:25(16):2078–2079.
https://doi.org/10.1093/bioinformatics/btp352.

Malinsky M, Matschiner M, Svardal H. Dsuite—fast D-statistics and related admixture evidence from VCF files. Mol Ecol
Resour. 2021:21(2):584–595. https://doi.org/10.1111/1755-0998.13265.

McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S,
Daly M, et al. The genome analysis toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res . 2010:20(9):1297–1303. https://doi.org/10.1101/gr.107524.110.

Mölder F, Jablonski K, Letcher B, Hall M, Tomkins-Tinch C, Sochat V, Forster J, Lee S, Twardziok S, Kanitz
A, et al. Sustainable data analysis with Snakemake [version 2; peer review: 2 approved]. F1000Res. 2021:10:33.
https://doi.org/10.12688/f1000research.29032.2.

Nadachowska-Brzyska K, Burri R, Smeds L, Ellegren H. PSMC analysis of effective population sizes in molecular
ecology and its application to black-and-white Ficedula flycatchers. Mol Ecol. 2016:25(5):1058–1072.
https://doi.org/10.1111/mec.13540.

Neph S, Kuehn MS, Reynolds AP, Haugen E, Thurman RE, Johnson AK, Rynes E, Maurano MT, Vierstra J,
Thomas S, et al. BEDOPS: high-performance genomic feature operations. Bioinformatics . 2012:28(14):1919–1920.
https://doi.org/10.1093/bioinformatics/bts277.

Patterson N, Moorjani P, Luo Y, Mallick S, Rohland N, Zhan Y, Genschoreck T, Webster T, Reich D. Ancient
admixture in human history. Genetics . 2012:192(3):1065–1093. https://doi.org/10.1534/genetics.112.145037.

Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D, Maller J, Sklar P, de Bakker PIW,
Daly MJ, et al. PLINK: a tool set for whole-genome association and population-based linkage analyses. Am J Hum Genet.
2007:81(3):559–575. https://doi.org/10.1086/519795.

Quinlan AR, Hall IM. BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics .
2010:26(6):841–842. https://doi.org/10.1093/bioinformatics/btq033.

Ramirez F, Ryan DP, Gruning B, Bhardwaj V, Kilpert F, Richter AS, Heyne S, Dundar F, Manke T. deepTools2:
a next generation web server for deep-sequencing data analysis. Nucleic Acids Res . 2016:44(W1):W160–W165.
https://doi.org/10.1093/nar/gkw257.

Rhie A, McCarthy SA, Fedrigo O, Damas J, Formenti G, Koren S, Uliano-Silva M, Chow W, Fungtammasan A,
Kim J. Towards complete and error-free genome assemblies of all vertebrate species. Nature . 2021:592(7856):737–746.
https://doi.org/10.1038/s41586-021-03451-0.

Schiffels S, Wang K. MSMC And MSMC2: the multiple sequentially markovian coalescent. In: Dutheil JY, editors. Statistical
population genomics . New York: Springer; 2020. p. 147–166.

Singhal S, Colston TJ, Grundler MR, Smith SA, Costa GC, Colli GR, Moritz C, Pyron RA, Rabosky DL.
Congruence and conflict in the higher-level phylogenetics of squamate reptiles: an expanded phylogenomic perspective. Syst Biol. 2021:70(3):542–557. https://doi.org/10.1093/sysbio/syaa054.

Smith SA, Dunn CW. Phyutility: a phyloinformatics tool for trees, alignments and molecular data. Bioinformatics .
2008:24(5):715–716. https://doi.org/10.1093/bioinformatics/btm619.

Smith SA, Moore MJ, Brown JW, Yang Y. Analysis of phylogenomic datasets reveals conflict, concordance, and gene
duplications with examples from animals and plants. BMC Evol Biol. 2015:15(1):150. https://doi.org/10.1186/s12862-015-0423-0.

Solís-Lemus C, Ané C. Inferring phylogenetic networks with maximum pseudolikelihood under incomplete lineage sorting.
PLoS Genet. 2016:12(3):e1005896. https://doi.org/10.1371/journal.pgen.1005896.

Solís-Lemus C, Bastide P, Ané C. PhyloNetworks: a package for phylogenetic networks. Mol Biol Evol.
2017:34(12):3292–3298. https://doi.org/10.1093/molbev/msx235.

Stamatakis A. RAxML version 8: a tool for phylogenetic analysis and post-analysis of large phylogenies. Bioinformatics .
2014:30(9):1312–1313. https://doi.org/10.1093/bioinformatics/btu033.

Sukumaran J, Holder MT. Dendropy: a python library for phylogenetic computing. Bioinformatics . 2010:26(12):1569–1571.
https://doi.org/10.1093/bioinformatics/btq228.

Wang K, Mathieson I, O'Connell J, Schiffels S. Tracking human population structure through time from whole genome
sequences. PLoS Genet. 2020:16(3):e1008552. https://doi.org/10.1371/journal.pgen.1008552.

Yan MY, Ferguson B, Bimber BN. VariantQC: a visual quality control report for variant evaluation. Bioinformatics .
2019:35(24):5370–5371. https://doi.org/10.1093/bioinformatics/btz560.

Yang ZH. PAML 4: phylogenetic analysis by maximum likelihood. Mol Biol Evol. 2007:24(8):1586–1591.
https://doi.org/10.1093/molbev/msm088.

Zhang C, Rabiee M, Sayyari E, Mirarab S. ASTRAL-III: polynomial time species tree reconstruction from partially
resolved gene trees. BMC Bioinformatics . 2018:19(S6):153. https://doi.org/10.1186/s12859-018-2129-y.
